cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/cache.go

```diff
@@ -0,0 +1,324 @@
+	watcherMutex sync.Mutex
+	redisCli     *redis.Client
+	clientGetter server.KubernetesClientGetter
+	// these waitGroup is used exclusively by unit tests to block until all expected repos have
+	// been indexed by the go routine running in the background. The creation of the WaitGroup object
+	// and .Add() is expected to be done by the unit test client. The server-side only signals
+	// when indexing one repo is complete
+	indexRepoWaitGroup *sync.WaitGroup
+}
+
+func newCache(clientGetter server.KubernetesClientGetter) (*fluxPlugInCache, error) {
+	log.Infof("+newCache")
+	REDIS_ADDR, ok := os.LookupEnv("REDIS_ADDR")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_ADDR")
+	}
+	REDIS_PASSWORD, ok := os.LookupEnv("REDIS_PASSWORD")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_PASSWORD")
+	}
+	REDIS_DB, ok := os.LookupEnv("REDIS_DB")
+	if !ok {
+		return nil, status.Errorf(codes.FailedPrecondition, "missing environment variable REDIS_DB")
+	}
+
+	REDIS_DB_NUM, err := strconv.Atoi(REDIS_DB)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Infof("newCache: addr: [%s], password: [%s], DB=[%d]", REDIS_ADDR, REDIS_PASSWORD, REDIS_DB)
+
+	return newCacheWithRedisClient(
+		clientGetter,
+		redis.NewClient(&redis.Options{
+			Addr:     REDIS_ADDR,
+			Password: REDIS_PASSWORD,
+			DB:       REDIS_DB_NUM,
+		}))
+}
+
+func newCacheWithRedisClient(clientGetter server.KubernetesClientGetter, redisCli *redis.Client) (*fluxPlugInCache, error) {
+	log.Infof("+newCacheWithRedisClient")
+
+	if redisCli == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "server not configured with redis Client")
+	}
+
+	if clientGetter == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "server not configured with configGetter")
+	}
+
+	// sanity check that the redis client is connected
+	pong, err := redisCli.Ping(redisCli.Context()).Result()
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("[PING] -> [%s]", pong)
+
+	c := fluxPlugInCache{
+		watcherStarted: false,
+		watcherMutex:   sync.Mutex{},
+		redisCli:       redisCli,
+		clientGetter:   clientGetter,
+	}
+	go c.startHelmRepositoryWatcher()
+	return &c, nil
+}
+
+func (c *fluxPlugInCache) startHelmRepositoryWatcher() {
+	log.Infof("+fluxv2 startHelmRepositoryWatcher")
+	c.watcherMutex.Lock()
+	// can't defer c.watcherMutex.Unlock() because when all is well,
+	// we never return from this func
+
+	if !c.watcherStarted {
+		ch, err := c.newHelmRepositoryWatcherChan()
+		if err != nil {
+			c.watcherMutex.Unlock()
+			log.Errorf("failed to start HelmRepository watcher due to: %v", err)
+			return
+		}
+		c.watcherStarted = true
+		c.watcherMutex.Unlock()
+		log.Infof("watcher successfully started. waiting for events...")
+
+		c.processEvents(ch)
+	} else {
+		c.watcherMutex.Unlock()
+		log.Infof("watcher already started. exiting...")
+	}
+	// we should never reach here
```

Comment:
Won't this point be reached if `startHelmRepositoryWatcher()` is called a second time? Ah, so I assume the assumption is that it is only ever called once, but in that case, I'm confused about the locks.... maybe it'll make more sense as I continue on :)

Reply:
yeah, I mean we should never reach here under normal conditions. if we reached there something bad happened. I will update the comment.